File: atm cajero sockets/cajero-cliente-go/src/atm/controlador.go
// Equivalente a atm/controlador.py: se inyecta cualquier implementación
// de ICajeroServidor (antes podría ser una local, ahora es el
// CajeroServidorProxy que habla con el banco por sockets). Al controlador
// no le importa.
package main

type CajeroControlador struct {
	servidor ICajeroServidor
}

func NewCajeroControlador(servidor ICajeroServidor) *CajeroControlador {
	return &CajeroControlador{servidor: servidor}
}

func (c *CajeroControlador) Depositar(numeroCuenta, clave string, monto float64) (bool, string, *float64) {
	return c.servidor.Depositar(numeroCuenta, clave, monto)
}

func (c *CajeroControlador) Retirar(numeroCuenta, clave string, monto float64) (bool, string, *float64) {
	return c.servidor.Retirar(numeroCuenta, clave, monto)
}

func (c *CajeroControlador) Consultar(numeroCuenta, clave string) (bool, string, *float64) {
	return c.servidor.Consultar(numeroCuenta, clave)
}

File: atm cajero sockets/cajero-cliente/src/atm/controlador.py
from typing import Optional, Tuple

from atm.interfaces import ICajeroServidor


class CajeroControlador:

    def __init__(self, servidor: ICajeroServidor):
        # Se inyecta una instancia que cumpla la interfaz ICajeroServidor.
        # Antes era un CajeroServidor local; ahora es un CajeroServidorProxy
        # que habla con el banco por sockets. Al controlador no le importa.
        self._servidor = servidor

    def depositar(self, numero_cuenta: str, clave: str, monto: float) -> Tuple[bool, str, Optional[float]]:
        return self._servidor.depositar(numero_cuenta, clave, monto)

    def retirar(self, numero_cuenta: str, clave: str, monto: float) -> Tuple[bool, str, Optional[float]]:
        return self._servidor.retirar(numero_cuenta, clave, monto)

    def consultar(self, numero_cuenta: str, clave: str) -> Tuple[bool, str, Optional[float]]:
        return self._servidor.consultar(numero_cuenta, clave)
